src/hardware/gpio.cpp

```diff
@@ -0,0 +1,79 @@
+#include "hardware/gpio.h"
+
+/**
+ * @brief Construct a new GPIO object
+ *
+ * Right now this is acting as an example peripheral, since reading/writing to
+ * GPIOs is atomic and doesn't require the use of synchronization primitives
+ * from the kernel.
+ *
+ * @param port The register block pointing to the GPIO port. (GPIOx)
+ * @param pin The pin *number*, not one of the macros provided by the HAL.
+ * @param mode Input/output/alternate as well as the push-pull/open drain mode.
+ * @param res Pull-up/pull-down resistor configuration.
+ * @param alt If `mode` is @ref GPIO::AlternatePP or @ref GPIO::AlternateOD,
+ * this selects the alternae function.
+ */
+GPIO::GPIO(GPIO::Port port, uint32_t pin, GPIO::Mode mode, GPIO::Resistor res,
+           uint32_t alt)
+    : port(port), regs(get_regs()), pin(1 << pin), mode(mode), res(res),
+      alt(alt) {}
+
+/**
+ * @brief Write to a GPIO pin.
+ *
+ * @param on true = HIGH, false = LOW
+ */
+void GPIO::set(bool on) { HAL_GPIO_WritePin(regs, pin, (GPIO_PinState)on); }
```

Comment:
Can we keep function names consist to what they are in the HAL, make this `GPIO::write`